contracts/MultiPartyEscrow.sol

```diff
@@ -0,0 +1,272 @@
+    //channel can be reused even after channel_claim(..., is_sendback=true)
+    function open_channel(address  recipient, uint256 value, uint256 expiration, uint256 replica_id) 
+    public
+    returns(bool) 
+    {
+        require(balances[msg.sender] >= value);
+        channels[next_channel_id] = PaymentChannel({
+            sender       : msg.sender,
+            recipient    : recipient,
+            value        : value,
+            replica_id   : replica_id,
+            nonce        : 0,
+            expiration   : expiration
+        });
+        balances[msg.sender] -= value;
+        emit event_channel_open(next_channel_id, msg.sender, recipient, replica_id);
+        next_channel_id += 1;
+        return true;
+    }
+    
+
+
+    function deposit_and_open_channel(address  recipient, uint256 value, uint256 expiration, uint256 replica_id)
+    public
+    returns(bool)
+    {
+        require(deposit(value));
+        require(open_channel(recipient, value, expiration, replica_id));
+        return true;
+    }
+
+    //open a channel from the recipient side. Sender should send the signed permission to open the channel
+    function open_channel_by_recipient(address  sender, uint256 value, uint256 expiration, uint256 replica_id, uint256 message_nonce, bytes memory signature) 
+    public
+    returns(bool) 
+    {
+        require(balances[sender] >= value);
+        require(isValidSignature_open_channel_replaysafe(msg.sender, value, expiration, replica_id, message_nonce, signature, sender));
+
+        channels[next_channel_id] = PaymentChannel({
+            sender       : sender,
+            recipient    : msg.sender,
+            value        : value,
+            replica_id   : replica_id,
+            nonce        : 0,
+            expiration   : expiration
+        });
+        balances[sender] -= value;
+
+        emit event_channel_open(next_channel_id, sender, msg.sender, replica_id);
+        next_channel_id += 1;
+        return true;
+    }
+ 
+    function _channel_sendback_and_reopen(uint256 channel_id)
+    private
+    {
+        PaymentChannel storage channel = channels[channel_id];
+        balances[channel.sender]      += channel.value; 
+        channel.value                  = 0;
+        channel.nonce                 += 1;
+        channel.expiration             = 0;
+    }
+
+    // the recipient can close the channel at any time by presenting a
+    // signed amount from the sender. The recipient will be sent that amount. The recipient can choose: 
+    // send the remainder to the sender (is_sendback == true), or put that amount into the new channel.
+    function channel_claim(uint256 channel_id, uint256 amount, bytes memory signature, bool is_sendback) 
+    public 
+    {
+        PaymentChannel storage channel = channels[channel_id];
+        require(amount <= channel.value);
+        require(msg.sender == channel.recipient);
```

Comment:
I do think in order to allow the service providers to stay "cold" as much as possible, having a check also on the actual `from` of the transaction limits the implementation of the daemon. So I would remove it

You can create in a row 10 signatures and then upload live on the daemon to use as "disposable" sigs.

Reply:
My understanding: in current design all snet-daemons (replicas) which share the same payment channel, should share the same recipient address as well, so this restriction should not prevent them from getting money. May be I didn't catch your point though.

Reply:
Marco.. No! In the current design it opens possibility for client to attack server ("close" the channel, but continue to use it).

Reply:
Marco.. No! In the current design it opens possibility for client to attack server ("close" the channel, but continue to use it).

Reply:
In future I'm planning to split the "signed" account from the "sending" account. Because we will need it for treasurer-server architecture in the case then server-by-something from the another server. But not now :)

Reply:
@astroseger, @tiero I am still being confused by your dialog because my understanding is what you can extract from signature is address of the client of the service (i.e. channel sender) but not the channel recipient's address.

Reply:
@astroseger, am I right that attack scenario you have mentioned is:
- client opens the channel
- server gets the channel id + nonce and consider this channel valid
- client calls ```channel_claim()``` with any (zero) amount to change channel nonce
- server doesn't know about channel state was changed, and verifies the payments containing the outdated channel nonce

Reply:
@vsbogd You are right, pardon I thought it was the consumer 👍